mkmamedb.man
.\" Converted with mdoc2man 0.2
.\" from NiH: mkmamedb.mdoc,v 1.7 2005/06/12 19:30:00 wiz Exp 
.\" $NiH: mkmamedb.mdoc,v 1.7 2005/06/12 19:30:00 wiz Exp $
.\"
.\" Copyright (c) 2005 Dieter Baron and Thomas Klausner.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above
.\"    copyright notice, this list of conditions and the following
.\"    disclaimer in the documentation and/or other materials provided
.\"    with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior
.\"    written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THOMAS KLAUSNER ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
.\" USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH MKMAMEDB 1 "July 1, 2005" NiH
.SH "NAME"
mkmamedb \- create database for use by ckmame
.SH "SYNOPSIS"
.B mkmamedb
[\fB-hV\fR]
[\fB-o\fR \fBdbfile\fR]
[\fB-prog-name\fR \fBname\fR]
[\fB-prog-version\fR \fBversion\fR]
[\fBrominfo-file\fR]
.SH "DESCRIPTION"
.B mkmamedb
parses a ROM set description, as provided by e.g.
xmame(6)
and converts it to an internal database format for use
by
ckmame(1).
The ROM set description is read from
\fBrominfo-file\fR
or stdin, if no
\fBrominfo-file\fR
is given on the command line.
The default output file is
\fBmame.db\fR.
You can override it with the
\fB-o\fR
option or the
\fBMAMEDB\fR
environment variable.
.PP
The following formats of ROM set description are supported: xmame
-listinfo, also known as CMPro; and xmame \-listxml.
The CMPro file format is described at
\fBhttp://www.logiqx.com/FAQs/CMPro.shtml\fR
the DTD for MAME can be found in
\fBPREFIX/share/doc/ckmame/\fR.
.SS "OPTIONS"
.B mkmamedb
supports the following options:
.RS
.TP 11
\fB-o\fR \fBdbfile, \fB--output\fR Ar dbfile\fR
Write to database
\fBdbfile\fR
instead of the default
``\fBmame.db\fR.''
.TP 11
\fB--prog-name\fR \fBname\fR
Set name of the program the ROM info is from.
.TP 11
\fB--prog-version\fR \fBversion\fR
Set version of the program the ROM info is from.
.TP 11
\fB-h\fR, \-help
Display a short help message.
.TP 11
\fB-V\fR, \-version
Display program name and version number.
.RE
.SH "ENVIRONMENT"
.RS
.TP 12
Ev MAMEDB
Location for the database file.
Can be overridden with the
\fB-o\fR
option.
Defaults to
``\fBmame.db\fR.''
.RE
.SH "EXAMPLES"
Get an input file for
mkmamedb(1)
from
xmame(6)
and convert it to the internal database format:
.Bd \-literal
# xmame \-listxml \*[Gt] db.txt
# mkmamedb db.txt
.Ed
.PP
Alternatively, if you have
xmame(6)
installed and in your
\fBPATH\fR,
you can just run
\fBmkmamedb-xmame.sh\fR.
.SH "SEE ALSO"
ckmame(1),
dumpgame(1)
.SH "AUTHORS"
-nosplit
.B mkmamedb
was written by
Dieter Baron <dillo@giga.or.at>
and
Thomas Klausner <tk@giga.or.at.>


NEWS
0.6 (2005-06-17)

- libzip is distributed separately
- Support MAME/MESS -listxml output
- Adapt for MAME -listinfo output changes (honor flags baddump and nodump).
- Check disk images (chd files).
- Superfluous merged into ckmame. ckmame now by default reports
  extra files.  Also works for samples.
- add file format version to database and don't use incompatible database
- dumpgame: find game by checksum of one of its ROMs
- add script to create database by running xmame

0.5 (2003-01-30)

MESS, Impact, Raine, and others based on the CMPro format now supported.
Man page added.
Bug fixes.

0.4.1

If a a clone zip didn't exist, it was not created.
The DOS port now really can fix romsets.

0.4

The whole program.


src/mkmamedb.c
/*
  $NiH: mkmamedb.c,v 1.30 2005/07/01 01:35:56 dillo Exp $

  mkmamedb.c -- create mamedb
  Copyright (C) 1999, 2003, 2004, 2005 Dieter Baron and Thomas Klausner

  This file is part of ckmame, a program to check rom sets for MAME.
  The authors can be contacted at <nih@giga.or.at>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License, version 2, as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/



#include <stdio.h>
#include <stdlib.h>

#include "config.h"

#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#else
#include "getopt.h"
#endif

#include "types.h"
#include "dbl.h"
#include "funcs.h"
#include "error.h"
#include "w.h"
#include "parse.h"

char *prg;
char *usage = "Usage: %s [-hV] [-o dbfile] [--prog-name name] [--prog-version version] [rominfo-file]\n";

char help_head[] = "mkmamedb (" PACKAGE ") by Dieter Baron and"
                   " Thomas Klausner\n\n";

char help[] = "\n\
  -h, --help               display this help message\n\
  -V, --version            display version number\n\
  -o, --output dbfile      write to database dbfile\n\
      --prog-name name     set name of program rominfo is from\n\
      --prog-version vers  set version of program rominfo is from\n\
\n\
Report bugs to <nih@giga.or.at>.\n";

char version_string[] = "mkmamedb (" PACKAGE " " VERSION ")\n\
Copyright (C) 2005 Dieter Baron and Thomas Klausner\n\
" PACKAGE " comes with ABSOLUTELY NO WARRANTY, to the extent permitted by law.\n\
You may redistribute copies of\n\
" PACKAGE " under the terms of the GNU General Public License.\n\
For more information about these matters, see the files named COPYING.\n";

#define OPTIONS "hVo:"

enum {
    OPT_PROG_NAME = 256,
    OPT_PROG_VERSION
};

struct option options[] = {
    { "help",          0, 0, 'h' },
    { "version",       0, 0, 'V' },
    { "output",        1, 0, 'o' },
    { "prog-name",     1, 0, OPT_PROG_NAME },
    { "prog-version",  1, 0, OPT_PROG_VERSION },
    { NULL,            0, 0, 0 },
};

int
main(int argc, char **argv)
{
    DB *db;
    char *dbname, *fname, *prog_name, *prog_version;
    int c;

    prg = argv[0];

    dbname = getenv("MAMEDB");
    if (dbname == NULL)
	dbname = DDB_DEFAULT_DB_NAME;
    fname = NULL;
    prog_name = prog_version = NULL;

    opterr = 0;
    while ((c=getopt_long(argc, argv, OPTIONS, options, 0)) != EOF) {
	switch (c) {
	case 'h':
	    fputs(help_head, stdout);
	    printf(usage, prg);
	    fputs(help, stdout);
	    exit(0);
	case 'V':
	    fputs(version_string, stdout);
	    exit(0);
	case 'o':
	    dbname = optarg;
	    break;
	case OPT_PROG_NAME:
	    prog_name = optarg;
	    break;
	case OPT_PROG_VERSION:
	    prog_version = optarg;
	    break;
    	default:
	    fprintf(stderr, usage, prg);
	    exit(1);
	}
    }

    switch (argc - optind) {
    case 1:
	fname = argv[optind];
	break;
    case 0:
	break;
    default:
	fprintf(stderr, usage, prg);
	exit(1);
    }
    
    remove(dbname);
    db = ddb_open(dbname, DDB_WRITE);

    if (db==NULL) {
	myerror(ERRDB, "can't create database '%s'", dbname);
	exit(1);
    }

    w_version(db);
    parse(db, fname, prog_name, prog_version);

    ddb_close(db);

    return 0;
}
